### PR TITLE
feat: Remove image refs in `defaults.go` and replace by env vars

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
Remove image refs in `defaults.go` and replace by env vars

### Reference issue
https://github.com/eclipse/che/issues/15874

### CODE REVIEW

File: pkg/controller/che/che_controller.go [hunk 1]
```diff
@@ -1254,3 +1257,41 @@ func hasConsolelinkObject() bool {
...
+
+	cheImageTag := util.GetValue(cr.Spec.Server.CheImageTag, deploy.DefaultCheVersion())
+
+	if len(cr.Spec.Server.CheImage) > 0 {
+		return cheImageRepo + ":" + cheImageTag
+	}
+
+	defaultCheServerImage := deploy.DefaultCheServerImage(cr)
+
+	if !strings.Contains(defaultCheServerImage, "@") {
```
Comment (davidfestal) [inline pkg/controller/che/che_controller.go]: I would apply the following logic even when the `defaultCheServerImage` (coming from the env variable) contains a digest.
In other words, if the `defaultCheServerImage` contains a digest, but the `cheImageTag` is overridden in the `CheCluster` custom resource, then I would replace either the tag or the digest of the `defaultCheServerImage` by the `CheCluster` `cheImageTag` value

**Comment by davidfestal** [comment]
Just had 2 comments when reviewing, but great work @AndrienkoAleksandr !

File: pkg/controller/che/che_controller.go (hunk 1 above)
Comment (davidfestal) [reply pkg/controller/che/che_controller.go]: @tolusha wdyt ?

File: release-operator-code.sh [hunk 2]
```diff
@@ -57,14 +66,25 @@ fi
...
+cat "${OPERATOR_YAML}" | \
+yq -ryY "( .spec.template.spec.containers[] | select(.name == \"che-operator\").env[] | select(.name == \"CHE_VERSION\") | .value ) = \"${RELEASE}\"" | \
+yq -ryY "( .spec.template.spec.containers[] | select(.name == \"che-operator\").env[] | select(.name == \"IMAGE_default_keycloak\") | .value ) = \"${KEYCLOAK_IMAGE_RELEASE}\"" | \
+yq -ryY "( .spec.template.spec.containers[] | select(.name == \"che-operator\").env[] | select(.name == \"IMAGE_default_plugin_registry\") | .value ) = \"${PLUGIN_REGISTRY_IMAGE_RELEASE}\"" | \
+yq -ryY "( .spec.template.spec.containers[] | select(.name == \"che-operator\").env[] | select(.name == \"IMAGE_default_devfile_registry\") | .value ) = \"${DEVFILE_REGISTRY_IMAGE_RELEASE}\"" \
+>> "${OPERATOR_YAML}.new"
+mv "${OPERATOR_YAML}.new" "${OPERATOR_YAML}"
 
 wget https://raw.githubusercontent.com/eclipse/che/${RELEASE}/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties -q -O /tmp/che.properties
 latestCheWorkspacePluginBrokerMetadataImage=$(cat /tmp/che.properties| grep "che.workspace.plugin_broker.metadata.image" | cut -d = -f2)
```
[the remark this answers is not in the document]
Comment (tolusha) [reply release-operator-code.sh]: From my understanding, the script does:
 - synchronizes `extra_images.go` is  with `che.properties`
 - updates `operator.yaml` by replacing a new image tag

File: pkg/controller/che/che_controller.go (hunk 1 above)
Comment (davidfestal) [reply pkg/controller/che/che_controller.go]: Thanks changing it @tolusha !

File: release-operator-code.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply release-operator-code.sh]: But are the variables defined in `extra_images.go` still used ?

[the remark this answers is not in the document]
Comment (davidfestal) [reply release-operator-code.sh]: Thanks for you change. That's really consistent now.

**Comment by nickboldt** [comment]
Is this planned for inclusion in 7.9?

**Comment by tolusha** [comment]
@nickboldt 
Yes

**Comment by tolusha** [comment]
[test]

**Comment by flacatus** [comment]
[test]

**Comment by nickboldt** [comment]
Looking forward to this feature being in Che and CRW.

Any advice for downstream implementation welcome, especially for files that have to be synced as is or modified for downstream.

https://github.com/redhat-developer/codeready-workspaces-operator/blob/master/operator.Jenkinsfile

**Comment by davidfestal** [comment]
So one last thing we seems to have forgotten: We should update the OLM packages in order to integrate the changes in the `operator.yaml`.

File: deploy/operator-local.yaml [hunk 3]
```diff
@@ -54,3 +54,24 @@ spec:
...
+            - name: IMAGE_default_devfile_registry
+              value: "quay.io/eclipse/che-devfile-registry:7.8.0"
+            - name: "IMAGE_default_pvc_jobs"
+              value: "registry.access.redhat.com/ubi8-minimal:8.0-213"
+            - name: "IMAGE_default_postgres"
+              value: "centos/postgresql-96-centos7:9.6"
+            - name: "IMAGE_default_keycloak"
+              value: "quay.io/eclipse/che-keycloak:7.8.0"
+            - name: "IMAGE_default_che_workspace_plugin_broker_metadata"
+              value: "quay.io/crw/pluginbroker-metadata-rhel8:2.1"
```
Comment (AndrienkoAleksandr) [inline deploy/operator-local.yaml]: Hmm... Guys, I see that our release scripts replace this images from crw to the che upstream:

`quay.io/crw/pluginbroker-metadata-rhel8:2.1 -> quay.io/eclipse/che-plugin-metadata-broker:v3.1.0`

 Maybe Should we do the same in this pr? Or Does our scripts mistake?

**Comment by AndrienkoAleksandr** [comment]
> So one last thing we seems to have forgotten: We should update the OLM packages in order to integrate the changes in the operator.yaml.
@tolusha @AndrienkoAleksandr

I think we can do that even after merge.

**Comment by davidfestal** [comment]
> > So one last thing we seems to have forgotten: We should update the OLM packages in order to integrate the changes in the operator.yaml.
> > @tolusha @AndrienkoAleksandr
> 
> I think we can do that even after merge.

well, that's not ideal. In any case we cannot afford merging this PR and release 7.9.0 while omitting the corresponding OLM change. That would break the next OLM version afaict

**Comment by AndrienkoAleksandr** [comment]
@davidfestal , @flacatus @tolusha Thank a lot for help and collaboration!

**Pull request merged**